Implementation for cats-effect 3

arixmkii opened the pull request:
Fixes #182 

Important version bumps:

- cats-core 2.4.2
- cats-effect 3.0.0-RC2
- scala 2.13.5
- prometheus 0.10.0

Additionally fixed some warnings generated by `source3` checks.

kubukoz commented on core/src/main/scala/io/chrisdavenport/epimetheus/Gauge.scala:
```diff
@@ -79,17 +79,17 @@ sealed abstract class Gauge[F[_]]{
   // Convenience
-  def incIn[E, F[_]: Bracket[?[_], E], A](g: Gauge[F], fa: F[A]): F[A] = 
-    Bracket[F, E].bracket(g.inc)(_ => fa)(_ => g.dec)
+  def incIn[E, F[_]: MonadCancel[*[_], E], A](g: Gauge[F], fa: F[A]): F[A] =
```
I think it's even a pattern in CE3 itself to use `MonadCancel[*[_], _]` in these situations (if E isn't part of the signature otherwise).

Also I think `*[_]` isn't supported by `-Ykind-projector` in dotty so might just want to pass an explicit implicit :)

arixmkii replied:
Hope I understood `explicit implicit` thing correctly here :)

kubukoz replied:
Ah, i meant passing the MonadCancel instamce using an `implicit` clause, as opposed to a context bound. Sorry for the confusion.

That way it'd be `MonadCancel[F, _]` I think.

arixmkii replied:
Yes, this is how I read this and moved `MonadCancel` to implicit parameter with the second type parameter unlocked.

kubukoz commented on core/src/main/scala/io/chrisdavenport/epimetheus/Gauge.scala:
```diff
@@ -203,7 +203,7 @@ object Gauge {
       case x: MapKUnlabelledGauge[_, _, _] => asJavaUnlabelled(x.base)
     }
-    def asJava[F[_]: ApplicativeError[?[_], Throwable]](c: Gauge[F]): F[JGauge] = c match {
+    def asJava[F[_]: ApplicativeError[*[_], Throwable]](c: Gauge[F]): F[JGauge] = c match {
```
You can use `ApplicativeThrow`

kubukoz commented on core/src/main/scala/io/chrisdavenport/epimetheus/Histogram.scala:
```diff
@@ -51,10 +48,10 @@ object Histogram {
-    Bracket[F, E].bracket(Clock[F].monotonic(unit))
-    {_: Long => fa}
-    {start: Long => Clock[F].monotonic(unit).flatMap(now => h.observe((now - start).toDouble))}
+  def timed[E, F[_]: MonadCancel[*[_], E]: Clock, A](h: Histogram[F], fa: F[A], unit: TimeUnit): F[A] =
```
Try `fa.timed`, it's in CE syntax: https://github.com/typelevel/cats-effect/pull/1677

kubukoz commented on core/src/test/scala/io/chrisdavenport/epimetheus/GaugeSpec.scala:
```diff
@@ -98,83 +97,75 @@ class GuageSpec extends Specification {
         defer <- Deferred[IO, Unit]
         fib <- gauge.incByIn(defer.get, 10).start
-        _ <- Timer[IO].sleep(1.second)
+        _ <- Temporal[IO].sleep(1.second)
```
Just `IO.sleep` would be good too btw

arixmkii commented:
- Bumped ce to 3.0-RC3;
- Added OpenMetrics support from prometheus `simple-client` version `0.10`;
- Used `.timed` from ce and relaxed `MonadCancel` requirement to just `FlatMap` for `Summary` and `Histogram`;
- Removed unnecessary instance summon in Spec file.

arixmkii commented:
@kubukoz  There is a difference between the old implementation of Timed using `bracket` (`MonadCancel`) and the one in ce3. The old implementation persists time metrics for failed (and I guess also for canceled) effect.

Example app:

```scala

import cats.effect.{ExitCode, IO, IOApp}
import io.chrisdavenport.epimetheus.{CollectorRegistry, Histogram, Name}

import scala.concurrent.duration._

object SummaryApp extends IOApp {

  override def run(args: List[String]): IO[ExitCode] =
    for {
      r <- CollectorRegistry.build[IO]
      hist <- Histogram.noLabels(r, Name("hist"), "help")
      _ <- Histogram.timedSeconds(hist, IO.sleep(4.3.seconds)
        .map(_ => throw new RuntimeException()))
        .handleErrorWith(_ => IO.unit)
      _ <- r.write004.map(println)
    } yield ExitCode.Success

}
```

This will have no data in registry for `timed` used from ce3.

I can see reasons to have measured successes only and all the effects, so, for me it is not clear which one is more valid than another. May be @ChristopherDavenport could comment and then I will update the scaladoc part.

Or may be we should expose both APIs for the user and may be even promote bracket version to ce3. 🤷

kubukoz commented:
Good catch, I didn't think of that. It's probably best to keep the timing method as it was, then.

arixmkii commented:
Reverted `timed` to their bracket version.

kubukoz commented:
Any blockers on this one?

ybasket commented:
Looks good to me – thank you @arixmkii 👍 

Hope it gets merged soon-ish, I came here because epimetheus (and -http4s) is one of my last blockers on CE3 adoption.

kubukoz commented:
Can we get a milestone with this? :)
